File: protocol/impls/go/ecmh.go
package main

// §13.2 — the pinned, portable ECMH primitive vector set (`sm ecmh`).
//
// Mirrors impls/c/src/ecmh.c's ecmh_cmd(): hash-to-curve KATs, identity
// serialization, a tagged multiset sum proven commutative, and an inverse
// round-trip — printed as a cross-language byte-identical `combined` golden
// (SPEC-conformance.md §13.2). Reuses this impl's self-rolled secp256k1.

import (
	"bytes"
	"crypto/sha256"
	"fmt"
)

// domain tags — second-preimage separation between tables.
const (
	ecmhTagName   = 0x01
	ecmhTagCommit = 0x02
	ecmhTagVote   = 0x03
	ecmhTagMut    = 0x04
	ecmhTagDecor  = 0x05
)

var ecmhRecTag = []byte{'E', 'C', 'M', 'H', 'v', '1'}

// ecmhFoldRow does acc ← acc + H2C("ECMHv1" ‖ tag ‖ rowBytes).
func ecmhFoldRow(acc []byte, tag byte, rowBytes []byte) {
	pre := append(append([]byte{}, ecmhRecTag...), tag)
	pre = append(pre, rowBytes...)
	pt, _ := secpEcmhHash(pre)
	secpEcmhAdd(acc, pt)
}

// stateEcmh is the incremental twin of stateDigest (§13.2): a per-table
// Elliptic-Curve Multiset Hash over the SAME canonical per-row encoding the
// state digest uses (so the two induce the identical equality relation),
// combined into one 32-byte value. A point-sum is order-independent, so a
// production fold maintains it in O(rows-changed)/block. Mirrors C
// sm_state_ecmh in impls/c/src/ecmh.c.
func (s *FoldState) stateEcmh() [32]byte {
	an := secpEcmhIdentity()
	ac := secpEcmhIdentity()
	av := secpEcmhIdentity()
	am := secpEcmhIdentity()
	ad := secpEcmhIdentity()

	// names — per-row fields BYTE-IDENTICAL to digest.go (owner_type NOT encoded).
	for k, r := range s.names {
		var b bytes.Buffer
		b.WriteByte(byte(len(k)))
		b.WriteString(k)
		b.Write(r.owner[:])
		b.WriteByte(byte(r.st))
		wI64(&b, r.leaseExpiry)
		b.Write(r.seller[:])
		b.WriteByte(r.sellerType)
		wU64(&b, r.price)
		wI64(&b, r.offerExpiry)
		b.Write(r.buyer[:])
		wU64(&b, r.burnLeg)
		wU64(&b, r.payLeg)
		wI64(&b, r.reserveExpiry)
		ecmhFoldRow(an, ecmhTagName, b.Bytes())
	}
	for _, c := range s.commits {
		var b bytes.Buffer
		b.Write(c.commitment[:])
		wI64(&b, c.height)
		wU32(&b, c.txIndex)
		wI64(&b, c.commitTime)
		ecmhFoldRow(ac, ecmhTagCommit, b.Bytes())
	}
	for _, v := range s.votes {
		var b bytes.Buffer
		b.Write(v.target[:])
		wU32(&b, v.vout)
		sc := v.score.bytesLE()
		b.Write(sc[:])
		ecmhFoldRow(av, ecmhTagVote, b.Bytes())
	}
	for o, h := range s.muts {
		var b bytes.Buffer
		b.Write(o[:])
		wI64(&b, h)
		ecmhFoldRow(am, ecmhTagMut, b.Bytes())
	}
	for _, d := range s.decors {
		var b bytes.Buffer
		b.Write(d.txid[:])
		wU32(&b, d.vout)
		b.WriteByte(byte(len(d.rec)))
		b.Write(d.rec)
		ecmhFoldRow(ad, ecmhTagDecor, b.Bytes())
	}

	// combined = SHA256("ECMHtop1" ‖ five sub-accumulators ‖ overflow flag).
	h := sha256.New()
	h.Write([]byte("ECMHtop1"))
	h.Write(an)
	h.Write(ac)
	h.Write(av)
	h.Write(am)
	h.Write(ad)
	if s.overflow {
		h.Write([]byte{1})
	} else {
		h.Write([]byte{0})
	}
	var out [32]byte
	copy(out[:], h.Sum(nil))
	return out
}

func runEcmh() {
	comb := sha256.New()
	feed := func(b []byte) { comb.Write(b) }

	// version self-doc
	fmt.Printf("ecmh ECMHv1\n")
	feed(ecmhRecTag)

	// 1. hash-to-curve KAT — fixed preimages → (ctr, compressed even-Y point).
	type h2cVec struct {
		label string
		pre   []byte
	}
	ff := bytes.Repeat([]byte{0xFF}, 32)
	z32 := make([]byte, 32)
	h2cs := []h2cVec{
		{"empty", []byte{}},
		{"a", []byte("a")},
		{"shib", []byte("shibpost")},
		{"doge", []byte("doge")},
		{"ff32", ff},
		{"z32", z32},
	}
	for _, v := range h2cs {
		pt, ctr := secpEcmhHash(v.pre)
		fmt.Printf("h2c %s ctr=%d pt=%s\n", v.label, ctr, hexstr(pt))
		feed([]byte{byte(ctr)})
		feed(pt)
	}

	// 2. identity (∞) serialization
	id := secpEcmhIdentity()
	fmt.Printf("identity %s\n", hexstr(id))
	feed(id)

	// 3. tagged multiset sum — a fixed set of (tag ‖ row) records, summed two ways.
	type rec struct {
		tag  byte
		body []byte
	}
	recs := []rec{
		{ecmhTagName, []byte("\x03foo")},
		{ecmhTagName, []byte("\x03bar")},
		{ecmhTagCommit, []byte("commitment-blob-32-bytes-xxxxxx")},
		{ecmhTagVote, []byte("vote-target-row")},
		{ecmhTagMut, []byte("owner-mutation")},
	}
	recPoint := func(r rec) []byte {
		pre := append(append([]byte{}, ecmhRecTag...), r.tag)
		pre = append(pre, r.body...)
		pt, _ := secpEcmhHash(pre)
		return pt
	}
	fwd := secpEcmhIdentity()
	for i := 0; i < len(recs); i++ {
		secpEcmhAdd(fwd, recPoint(recs[i]))
	}
	rev := secpEcmhIdentity()
	for i := len(recs) - 1; i >= 0; i-- {
		secpEcmhAdd(rev, recPoint(recs[i]))
	}
	commut := 0
	if bytes.Equal(fwd, rev) {
		commut = 1
	}
	fmt.Printf("sum %s\n", hexstr(fwd))
	fmt.Printf("commutative %d\n", commut)
	feed(fwd)
	feed([]byte{byte(commut)})

	// 4. inverse — remove the first record from the sum, re-add, must round-trip.
	pt0 := recPoint(recs[0])
	acc := append([]byte{}, fwd...)
	npt := append([]byte{}, pt0...)
	secpEcmhNegate(npt)
	secpEcmhAdd(acc, npt) // remove rec[0]
	secpEcmhAdd(acc, pt0) // re-add rec[0]
	roundtrip := 0
	if bytes.Equal(acc, fwd) {
		roundtrip = 1
	}
	fmt.Printf("inverse_roundtrip %d\n", roundtrip)
	feed([]byte{byte(roundtrip)})

	cd := comb.Sum(nil)
	fmt.Printf("combined %s\n", hexstr(cd))
}

File: protocol/impls/rust/src/ecmh.rs
//! §13.2 — the pinned, portable ECMH primitive vector set (`sm ecmh`).
//!
//! Faithful port of impls/c/src/ecmh.c (`ecmh_cmd`). Prints output byte-identical
//! to the C reference: hash-to-curve KATs over fixed preimages, identity (∞)
//! serialization, a tagged multiset sum (asserted commutative + pinned), and an
//! inverse round-trip — all folded into one cross-language `combined` digest.

use crate::secp256k1::{secp_ecmh_add, secp_ecmh_hash, secp_ecmh_identity, secp_ecmh_negate};
use crate::sha256::Sha256;

const HEXD: &[u8; 16] = b"0123456789abcdef";
fn puthex(out: &mut String, d: &[u8]) {
    for &b in d {
        out.push(HEXD[(b >> 4) as usize] as char);
        out.push(HEXD[(b & 15) as usize] as char);
    }
}

// domain tags — second-preimage separation between tables.
const TAG_NAME: u8 = 0x01;
const TAG_COMMIT: u8 = 0x02;
const TAG_VOTE: u8 = 0x03;
const TAG_MUT: u8 = 0x04;
const ECMH_REC_TAG: [u8; 6] = [b'E', b'C', b'M', b'H', b'v', b'1'];

// P(rec) = H2C("ECMHv1" ‖ tag ‖ body).
fn rec_point(tag: u8, body: &[u8]) -> [u8; 33] {
    let mut pre = Vec::with_capacity(7 + body.len());
    pre.extend_from_slice(&ECMH_REC_TAG);
    pre.push(tag);
    pre.extend_from_slice(body);
    let (_ctr, pt) = secp_ecmh_hash(&pre);
    pt
}

pub fn run() -> i32 {
    let mut comb = Sha256::new();
    let mut line = String::new();

    // version self-doc
    println!("ecmh ECMHv1");
    comb.update(b"ECMHv1");

    // 1. hash-to-curve KAT — fixed preimages → (ctr, compressed even-Y point).
    let h2c: [(&str, &[u8]); 4] = [
        ("empty", b""),
        ("a", b"a"),
        ("shib", b"shibpost"),
        ("doge", b"doge"),
    ];
    let ff = [0xFFu8; 32];
    let z32 = [0u8; 32];
    let mut kat: Vec<(&str, &[u8])> = h2c.to_vec();
    kat.push(("ff32", &ff));
    kat.push(("z32", &z32));
    for (label, pre) in &kat {
        let (ctr, pt) = secp_ecmh_hash(pre);
        line.clear();
        puthex(&mut line, &pt);
        println!("h2c {} ctr={} pt={}", label, ctr, line);
        comb.update(&[ctr as u8]);
        comb.update(&pt);
    }

    // 2. identity (∞) serialization
    let id = secp_ecmh_identity();
    line.clear();
    puthex(&mut line, &id);
    println!("identity {}", line);
    comb.update(&id);

    // 3. tagged multiset sum — a fixed set of (tag ‖ row) records, summed two ways.
    let recs: [(u8, &[u8]); 5] = [
        (TAG_NAME, b"\x03foo"),
        (TAG_NAME, b"\x03bar"),
        (TAG_COMMIT, b"commitment-blob-32-bytes-xxxxxx"),
        (TAG_VOTE, b"vote-target-row"),
        (TAG_MUT, b"owner-mutation"),
    ];
    let mut fwd = secp_ecmh_identity();
    let mut rev = secp_ecmh_identity();
    for (tag, body) in recs.iter() {
        let pt = rec_point(*tag, body);
        secp_ecmh_add(&mut fwd, &pt);
    }
    for (tag, body) in recs.iter().rev() {
        let pt = rec_point(*tag, body);
        secp_ecmh_add(&mut rev, &pt);
    }
    let commut = (fwd == rev) as u8;
    line.clear();
    puthex(&mut line, &fwd);
    println!("sum {}", line);
    println!("commutative {}", commut);
    comb.update(&fwd);
    comb.update(&[commut]);

    // 4. inverse — remove the first record from the sum, re-add, must round-trip.
    {
        let pt0 = rec_point(recs[0].0, recs[0].1);
        let mut acc = fwd;
        let mut npt = pt0;
        secp_ecmh_negate(&mut npt);
        secp_ecmh_add(&mut acc, &npt); // remove rec[0]
        secp_ecmh_add(&mut acc, &pt0); // re-add rec[0]
        let roundtrip = (acc == fwd) as u8;
        println!("inverse_roundtrip {}", roundtrip);
        comb.update(&[roundtrip]);
    }

    let cd = comb.finalize();
    line.clear();
    puthex(&mut line, &cd);
    println!("combined {}", line);
    0
}
